UX improved & fixes added

TheHussky opened the pull request:
1)Requested UX changes done
2)Requested comments added
3)Minor fixes added

TheHussky committed:
UX improved & fixes added

mmskv commented on src/dexpo.cpp:
```diff
@@ -77,27 +76,37 @@ main ()
           {
             w.draw_gui ();
-            // w.draw_border (0, dexpo_hlcolor);
+            for (int i = 0; int (i < w.desktops.size ()); i++)
+              {
+                w.draw_border (i, dexpo_desktop_color);
+              }
```
lmao why would you declare `i` as an `int` and then explicitly convert it to compare to `size_t` both in `draw_border` and in `int (i < w.desktops.size ()`

Declaring `i` as `size_t` would solve both these issues
though I would change all `size_t` to `uint`

TheHussky replied:
Declaring i as size_t would solve both these issues
Noticed that after push, will be solved.

mmskv commented on src/dexpo.cpp:
```diff
@@ -77,27 +76,37 @@ main ()
+              {
+                w.draw_border (i, dexpo_desktop_color);
+              }
+            w.desktop_sel = get_property_value (
+                window::c_, window::screen_->root, "_NET_CURRENT_DESKTOP")[0];
```
bro there is a function `get_current_desktop` in `xcb_utils.cpp`

TheHussky replied:
fuck

mmskv commented on src/dexpo.cpp:
```diff
@@ -77,27 +76,37 @@ main ()
             w.clear_preselection ();
-            if (next) // TODO(mangalinor) Document code
+            if (next)
+              // Pressing right arrow or down arrow
+              // Preselects next desktop
```
That's too explicit, will change it myself

mmskv commented on src/dexpo.cpp:
```diff
@@ -106,6 +115,14 @@ main ()
               {
                 ewmh_change_desktop (window::c_, window::screen_,
                                      w.desktop_sel);
+                // Next line fixes bug that appears while switching to active
+                // desktop
+                if (w.desktop_sel
+                    == get_property_value (window::c_, window::screen_->root,
+                                           "_NET_CURRENT_DESKTOP")[0])
+                  {
+                    return 0;
+                  };
```
ewmh_change_desktop should not run if `w.desktop_sel == get_current_desktop`

What bug does this ifx?

TheHussky replied:
1. Adding return 0 or exit(0) to cases of left mouse click and enter press crashes them - the desktop doesn't switch.(Idk why tho, maybe exit happens faster than x server reply or smth like that)
2. (1) is fixed the following way: exit is done in focus out. But choosing the desktop from which dexpo was launched doesn't make focus lost. So it requires special case.

TheHussky replied:
Do you mean emwh change desktop would fit better in else?

mmskv replied:
> The desktop doesn't switch.

You need to do `xcb_flush(c)`


> Do you mean emwh change desktop would fit better in else?

Now that i think about it, running ewmh_change_desktop on current desktop won't do anything, so leave it as it is (remove lines 120-122)

mmskv commented on src/dexpo.cpp:
```diff
@@ -127,37 +144,25 @@ main ()
         case XCB_FOCUS_OUT:
-          // TODO(mangalinor) Document code
-          // (what is the difference between XCB_LEAVE_NOTIFY and XCB_FOCUS_OUT)
+          // This happens when desktop is switched or
+          // mouse button is pressed in another window
```
More explicit documentation would fit better
like: "Quit dexpo on lost focus"

mmskv replied:
No need to tell that XCB_FOCUS_OUT happens when application loses focus. This is straightforward

TheHussky replied:
Will fix

mmskv commented on src/config.hpp:
```diff
@@ -17,9 +17,10 @@ const int dexpo_height = 0;
-const uint32_t dexpo_bgcolor = 0xAA525252; ///< Background color. TODO Alpha
+const uint16_t dexpo_hlwidth = 4; ///< Width of the preselected desktop border
+const uint32_t dexpo_bgcolor = 0xAA525252; ///< Background color. TODO Alpha and document properly
+const uint32_t dexpo_preselected_desktop_color = 0Xc20a66;///< Color of the preselection border
+const uint32_t dexpo_desktop_color =   0X0066cc;///<Color of the unselected desktops borders
```
Use clang autoformatter god damn

TheHussky replied:
I did? God knows wth happened

mmskv reviewed (changes requested):
Minor changes required

TheHussky committed:
Requsted changes

TheHussky reviewed (commented):
lgtm

mmskv commented on src/config.hpp:
```diff
@@ -20,7 +20,7 @@ const uint16_t dexpo_padding = 5; ///< Padding around the screenshots
 const uint32_t dexpo_bgcolor = 0xAA525252; ///< Background color. TODO Alpha and document properly
 const uint32_t dexpo_preselected_desktop_color = 0Xc20a66;///< Color of the preselection border
-const uint32_t dexpo_desktop_color =   0X0066cc;///<Color of the unselected desktops borders
+const uint32_t dexpo_desktop_color =   0X0066cc;///< Color of the unselected desktops borders
```
Well this is clearly not formatted
too many spaces: ```
... =   0X0066cc...```

No space between code and comment:
```cc;///< Color o...```

mmskv replied:
I will fix this in the next commit

mmskv committed:
Replace uint16_t with uint in drawables

mmskv committed:
Refactor: readability, docs, remove useless casts

- Replace size_t with uint
- Removed useless casts. Implicit conversions won't lose data I think
- Forgot to add drawable.cpp to previous commit

mmskv committed:
Rename config variables. Add sensible defaults

mmskv commented:
Border size should not conflict with padding size.

For example setting `dexpo_border_pres_width` to values greater than `dexpo_padding` should not remove padding

mmskv committed:
Do not obstruct padding with borders

- Rename some functions

---
Not sure how to document all this math

mmskv committed:
Change sensible defaults to be more sensible

---
Very subjective

mmskv committed:
Enable avoid-c-arrays check

TheHussky commented on src/desktop.cpp:
```diff
@@ -27,15 +27,15 @@ dxp_desktop::dxp_desktop (
   this->pixmap_width = dexpo_width;
   this->pixmap_height = dexpo_height;
 
+  float screen_ratio = float (this->width) / this->height; ///< width/height
```
Do you think this comment is even needed?

mmskv replied:
In this case not really, as the variable scope is very small.

screen_ratio doesn't convey whether it is width/height or height/width. As the comment is in the doxygen format it can be parsed by Language Server and shown to programmer when he hovers over variable
